cfme/tests/middleware/test_middleware_datasource.py

```diff
@@ -122,3 +123,22 @@ def _get_datasources_set(datasources):
     such as 'name' and 'server'
     """
     return set((datasource.name, datasource.server.name) for datasource in datasources)
+
+
+def test_tags(provider):
+    """Tests tags in datasources page
+
+    Steps:
+        * Select a datasource randomly from database
+        * Run `validate_tags` with `tags` input
+    """
+    tags = [
+        Tag(category=Category(display_name='Environment', single_value=True), display_name='Test'),
+        Tag(category=Category(display_name='Location', single_value=True), display_name='New York'),
+        Tag(category=Category(display_name='Workload', single_value=False),
+            display_name='Application Servers'),
+    ]
+    ds_db = MiddlewareDatasource.datasources_in_db(provider=provider)
+    assert len(ds_db) > 0, "There is no datasource(s) available in UI"
+    datasource = get_random_list(ds_db, 1)[0]
+    datasource.validate_tags(tags=tags)
```

Comment:
OK, this method should **definitely** be renamed. It doesn't validate. It runs an entire check and obliterates tags that may already be assigned. What should happen here is that the tags are added as part of a fixture and validate here as part of the test, assuming that adding the tests is not part of the test, if it is, a separate add, and then a separate validate and delete should exist. Like a crud test. @mfalesni @jkrocil agree?